src/nbl/asset/utils/CSPIRVIntrospector.cpp

```diff
@@ -13,472 +13,528 @@ namespace nbl::asset
-
-            const auto& introBinding = introBindings[checkedDescsCnt[i]].binding;
-            if (introBinding < binding.binding)
-                binding.binding = introBinding;
-        }
-        if (!anyStageNotFinished) //all shader stages finished
-            break;
+    SPushConstantRange current = {
+        .offset = 0,
+        .size = 0
+    };
+
+    // TODO: test
+    // run-length encode m_pushConstantBytes
+    for (uint32_t currentByteOffset = 1u; currentByteOffset < MaxPushConstantsSize; ++currentByteOffset)
+    {
+        current.stageFlags = m_pushConstantBytes[currentByteOffset].value;
+        current.offset++;
+        current.size++;
 
-        const ICPUSpecializedShader* refShader = nullptr;
-        uint32_t refIndex = ~0u;
-        core::smart_refctd_ptr<const CIntrospectionData> refIntro = nullptr;
+        if (current.stageFlags != prev.stageFlags)
         {
-            auto checkedDescsCntIt = checkedDescsCnt;
-            auto introspectionIt = introspections;
-            for (auto shader : shaders)
+            if (prev.stageFlags)
             {
-                auto& introBindings = (*introspectionIt)->descriptorSetBindings[_set];
-                if (*checkedDescsCntIt!=introBindings.size())
-                {
-                    const auto& introBinding = introBindings[*checkedDescsCntIt].binding;
-                    if (introBinding==binding.binding)
-                    {
-                        stageFlags |= shader->getStage();
-                        refIndex = (*checkedDescsCntIt)++;
-                        refIntro = *introspectionIt;
-                        refShader = shader;
-                    }
-                }
-                introspectionIt++;
-                checkedDescsCntIt++;
+                prev.size = current.offset - prev.offset;
+                tmp.push_back(prev);
             }
+            prev = current;
         }
+    }
 
-        auto& introBnd = refIntro->descriptorSetBindings[_set][refIndex];
-        binding.type = resType2descType(introBnd.type);
-        binding.stageFlags = static_cast<ICPUShader::E_SHADER_STAGE>(stageFlags);
-        binding.count = introBnd.descriptorCount;
-        if (introBnd.descCountIsSpecConstant)
-        {
-            auto& specInfo = refShader->getSpecializationInfo();
-            auto val = specInfo.getSpecializationByteValue(binding.count);
-            assert(val.second == 4ull);
-            memcpy(&binding.count, val.first, 4ull);
-        }
+    if (prev.stageFlags)
+    {
+        prev.size = current.offset - prev.offset;
+        tmp.push_back(prev);
+    }
+
+    if(tmp.size() == 0)
+        return nullptr;
 
-        bindings.push_back(binding);
+    std::ostringstream debug;
+    for (auto it = tmp.begin(); it != tmp.end(); it++)
+        debug << "Stage flags: " << std::bitset<32>(it->stageFlags) << " offset: " << it->offset << " size: " << it->size << std::endl;
+    std::cout << debug.str();
```

Comment:
would be cool to use a logger for this and make a `CPipelineIntrospectionData::debugPrint(system::ILogger* logger) const` method for all this

Reply:
`debugPrint` as a distinct function?